packages/wordcount/src/defaultSettings.js

```diff
@@ -1,3 +1,40 @@
+/** @typedef {import('./index').WPWordCountStrategy} WPWordCountStrategy */
+
+/** @typedef {Partial<{type: WPWordCountStrategy, shortcodes: string[]}>} WPWordCountL10n */
+
+/**
+ * @typedef WPWordCountSettingsFields
+ * @property {RegExp}              HTMLRegExp                        Regular expression that matches HTML tags
+ * @property {RegExp}              HTMLcommentRegExp                 Regular expression that matches HTML comments
+ * @property {RegExp}              spaceRegExp                       Regular expression that matches spaces in HTML
+ * @property {RegExp}              HTMLEntityRegExp                  Regular expression that matches HTML entities
+ * @property {RegExp}              connectorRegExp                   Regular expression that matches word connectors, like em-dash
+ * @property {RegExp}              removeRegExp                      Regular expression that matches various characters to be removed when counting
+ * @property {RegExp}              astralRegExp                      Regular expression that matches astral UTF-16 code points
+ * @property {RegExp}              wordsRegExp                       Regular expression that matches words
+ * @property {RegExp}              characters_excluding_spacesRegExp Regular expression that matches characters excluding spaces
+ * @property {RegExp}              characters_including_spacesRegExp Regular expression that matches characters including spaces
+ * @property {RegExp}              shortcodesRegExp                  Regular expression that matches WordPress shortcodes
+ * @property {string[]}            shortcodes                        List of all shortcodes
+ * @property {WPWordCountStrategy} type                              Describes what and how are we counting
+ * @property {WPWordCountL10n}     l10n                              Object with human translations
+ */
+
+// Disable reason: JSDoc linter doesn't seem to parse the union (`&`) correctly: https://github.com/jsdoc/jsdoc/issues/1285
+/* eslint-disable jsdoc/valid-types */
+/**
+ * Lower-level settings for word counting that can be overridden.
+ *
+ * @typedef {Partial<WPWordCountSettingsFields>} WPWordCountUserSettings
+ */
+
+/**
+ * Word counting settings that include non-optional values we set if missing
+ *
+ * @typedef {WPWordCountUserSettings & typeof defaultSettings} WPWordCountDefaultSettings
+ */
+/* eslint-enable jsdoc/valid-types */
```

Comment:
I would like to avoid this type completely. I believe we can get to a place where we work with a fairly strict interface that's already defined as `WPWordCountSettingsFields`. `loadSettings` should be able to return a well typed interface on its inference of `defaultSettings` and `userSettings`.

Reply:
Hm, I agree on principle that it would be nice if we had one less type, @sirreal. At the same time, if we want to commit to always returning a full interface we will need to commit to firmer decisions around `l10n` and `shortcodes` – their goal is bluerrier to me at this point and the risk outweighs my desire for purity and simplicity at this point. If you’d like to take a stab at it I’d support you with a review :)

Reply:
This was part of what I did in the diff I shared [here](https://github.com/WordPress/gutenberg/pull/22077#issuecomment-634164997), but I'm happy to drop that for now.